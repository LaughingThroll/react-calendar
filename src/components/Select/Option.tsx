import React from "react"

interface IOption extends React.OptionHTMLAttributes<HTMLOptionElement> {
  title: string
}

const Option: React.FC<IOption> = ({ title, ...rest }) => {
  return (
    <option className="select__option" {...rest}>
      {title}
    </option>
  )
}

export default Option
