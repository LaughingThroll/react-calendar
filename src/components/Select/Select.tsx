import React from "react"

interface ISelect extends React.SelectHTMLAttributes<HTMLSelectElement> {
  onChange: (e: React.ChangeEvent<HTMLSelectElement>) => void
  children?: JSX.Element | JSX.Element[]
  value?: string
}

const Select: React.FC<ISelect> = ({ children, value, ...rest }) => {
  return (
    <div className="select">
      <select className="select__main" value={value} {...rest}>
        {children}
      </select>
    </div>
  )
}

export default Select
