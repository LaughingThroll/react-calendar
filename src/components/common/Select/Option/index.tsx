import React from 'react'
import styles from './option.module.scss'

export interface OptionProps extends React.OptionHTMLAttributes<HTMLOptionElement> {
  title: string
}

const Option: React.FC<OptionProps> = ({ title, ...rest }) => {
  const { select__option } = styles
  return (
    <option className={select__option} {...rest}>
      {title}
    </option>
  )
}

export default React.memo(Option)
