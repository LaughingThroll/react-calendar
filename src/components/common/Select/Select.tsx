import React from 'react'

interface ISelect extends React.SelectHTMLAttributes<HTMLSelectElement> {
  onChange: (e: React.ChangeEvent<HTMLSelectElement>) => void
  children?: JSX.Element | JSX.Element[]
  value?: string | number
}

const Select: React.FC<ISelect> = ({ children, value, ...rest }) => {
  return (
    <div className="select">
      <select className="select__main" value={value} {...rest}>
        {children}
      </select>
    </div>
  )
}

export default React.memo(Select)
