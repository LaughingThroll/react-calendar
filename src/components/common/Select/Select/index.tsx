import React from 'react'
import styles from './select.module.scss'

export interface SelectProps extends React.SelectHTMLAttributes<HTMLSelectElement> {
  onChange: (e: React.ChangeEvent<HTMLSelectElement>) => void
  value: string | number
  children?: any
}

const isOption = (child: any): boolean => child.type === 'option' || child.type?.type?.name === 'Option'
const renderOptions = (child: any): any | never => {
  try {
    if (isOption(child)) return child
  } catch (error) {
    throw new Error(error)
  }
}

const Select: React.FC<SelectProps> = ({ children, value, ...rest }) => {
  const { select, select__main } = styles

  return (
    <div className={select}>
      <select className={select__main} value={value} {...rest}>
        {React.Children.map(children, renderOptions)}
      </select>
    </div>
  )
}

export default React.memo(Select)
