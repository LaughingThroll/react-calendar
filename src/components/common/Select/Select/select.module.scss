@import 'styles/utils/variables';

.select {
  position: relative;
  &::before {
    content: '';
    width: 12px;
    height: 5px;
    position: absolute;
    top: 50%;
    right: 18px;
    transform: translate(0, -50%);
    background-position: center center;
    background-size: cover;
    background-repeat: no-repeat;
    background-image: url('../../../../assets/images/icons/arrow--down.svg');
    cursor: pointer;
  }
  &__main {
    border-radius: 3px;
    background-color: $select-bg;
    width: 100%;
    padding: 12px 18px 11px 17px;
    appearance: none;
    outline: none;
    cursor: pointer;
  }
}
