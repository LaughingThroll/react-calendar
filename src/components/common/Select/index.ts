export { default as Select } from './Select/index'
export { default as Option } from './Option/index'
