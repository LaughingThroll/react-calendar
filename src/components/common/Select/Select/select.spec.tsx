import React from 'react'
import { shallow, render } from 'enzyme'
import Select, { SelectProps } from './'

const setUp = (props: SelectProps) => shallow(<Select {...props} />)

describe('Select render', () => {
  it('default render', () => {
    const tree = setUp({ onChange: jest.fn(), value: 1 })

    expect(tree).toMatchSnapshot()
    expect(tree.find('.select')).toHaveLength(1)
    expect(tree.find('select')).toHaveLength(1)
    expect(tree.find('select').prop('value')).toEqual(1)
  })
})

describe('Select props', () => {
  it('props onChange', () => {
    const props = { onChange: jest.fn(), value: 1 }
    const tree = setUp(props)
    const $input = tree.find('select')

    $input.simulate('change')

    expect(props.onChange).toHaveBeenCalled()
  })

  it('props value number', () => {
    const props = { onChange: jest.fn(), value: 1 }
    const tree = setUp(props)
    expect(tree.find('select').prop('value')).toEqual(1)
  })

  it('props value string', () => {
    const props = { onChange: jest.fn(), value: 'string' }
    const tree = setUp(props)
    expect(tree.find('select').prop('value')).toEqual('string')
  })

  it('props children should be exception', () => {
    const props = { onChange: jest.fn(), value: 1 }

    const tree = shallow(
      <Select {...props}>
        <div>ds</div>
      </Select>
    )
    expect(() => tree.simulateError(Error)).toThrow()
  })

  it('props children with options', () => {
    const props = { onChange: jest.fn(), value: 1 }
    const tree = render(
      <Select {...props}>
        <option>ds</option>
      </Select>
    )

    expect(tree.find('option')).toHaveLength(1)
  })
})
