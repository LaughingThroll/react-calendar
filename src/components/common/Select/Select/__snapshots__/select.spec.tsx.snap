// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select render default render 1`] = `
<div
  className="select"
>
  <select
    className="select__main"
    onChange={[MockFunction]}
    value={1}
  />
</div>
`;
